src/infra/core/src/transform_service_impl.rs

```diff
@@ -374,13 +366,30 @@ impl TransformServiceImpl {
 
         // Determine unprocessed block and offset range
         let last_unprocessed_block = input_chain.resolve_ref(&BlockRef::Head).await.int_err()?;
+
+        type Flag = MetadataEventTypeFlags;
+        type Decision = MetadataVisitorDecision;
+
         let last_unprocessed_offset = input_chain
-            .iter_blocks_interval(&last_unprocessed_block, last_processed_block, false)
-            .filter_map_ok(|(_, b)| b.into_data_stream_block())
-            .try_next()
-            .await
-            .int_err()?
-            .and_then(|b| b.event.last_offset())
+            .reduce_by_hash(
+                &last_unprocessed_block,
+                None,
+                Decision::NextOfType(Flag::DATA_BLOCK),
+                |state, hash, block| {
+                    if last_processed_block == Some(hash) {
+                        return Ok(Decision::Stop);
+                    }
+
+                    let Some(data_block) = block.as_data_stream_block() else {
```

Comment:
ditto

Reply:
I think you might like the new version better -- Done